apps/system/js/sync_manager.js

```diff
@@ -0,0 +1,460 @@
+      this.debug('onsyncenabling observed');
+      this.setState();
+
+      this.getAssertion().then(() => {
+        window.addEventListener(FXA_EVENT, this._handle_fxaEvent);
+        Service.request('SyncStateMachine:success');
+      }).catch(error => {
+        console.error('Could not enable sync', error);
+        Service.request('SyncStateMachine:error', ERROR_GET_FXA_ASSERTION);
+      });
+    },
+
+    _handle_onsyncerrored: function(event) {
+      this.debug('onsyncerrored observed');
+
+      this.setState();
+
+      var error = event.detail && Array.isArray(event.detail.args) ?
+                  event.detail.args[0] : null;
+      if (!error) {
+        return;
+      }
+
+      this.setError(error);
+
+      if (error in SyncUnrecoverableErrors) {
+        Service.request('SyncStateMachine:disable');
+      }
+    },
+
+    _handle_onsyncsyncing: function() {
+      this.debug('onsyncsyncing observed');
+      this.setState();
+
+      // In case that we are syncing on user demand, we unregister the
+      // sync request.
+      // It's harmless to unregister an expired request, so it's also ok
+      // to do it for a periodic sync.
+      this.unregisterSyncRequest();
+
+      if (!this.collections.size) {
+        Service.request('SyncStateMachine:success');
+        return;
+      }
+
+      var args = [];
+      this.getAssertion().then(assertion => {
+        args.push(assertion);
+        return this.getKeys();
+      }).then(keys => {
+        args.push(keys);
+        args.push([...this.collections.keys()]);
+        this.doSync.apply(this, args);
+      }).catch(error => {
+        console.error('Sync error', error);
+        Service.request('SyncStateMachine:error', ERROR_SYNC_REQUEST);
+      });
+    },
+
+    _handle_fxaEvent: function(event) {
+      if (!event || !event.detail) {
+        console.error('Wrong event');
+        return;
+      }
+
+      if (event.detail.eventName != 'onlogout') {
+        return;
+      }
+
+      // Logging out from Firefox Accounts should disable Sync.
+      Service.request('SyncStateMachine:disable');
+    },
+
+    _handle_killapp: function(event) {
+      // The synchronizer app can be killed while it's handling a sync
+      // request. In that case, we need to record an error and notify
+      // the state machine about it. Otherwise we could end up on a
+      // permanent 'syncing' state.
+      var killedApp = event.detail.origin;
+      navigator.mozApps.getSelf().onsuccess = event => {
+        var app = event.target.result;
+        app.getConnections().then(connections => {
+          connections.forEach(connection => {
+            if (connection.keyword != SYNC_REQUEST_IAC_KEYWORD ||
+                connection.subscriber != killedApp) {
+              return;
+            }
+            Service.request('SyncStateMachine:error', ERROR_SYNC_APP_KILLED);
+          });
+
+        });
+      };
+    },
+
+    /** Settings observers **/
```

Comment:
Yes, I can add a common handler helper, but unfortunately the handlers themselves need to be split. That's how BaseModule defines it.